src/sfloader/fluid_sfont.c

```diff
@@ -546,3 +551,242 @@ int fluid_sample_set_pitch(fluid_sample_t* sample, int root_key, int fine_tune)
+    {
+        FLUID_LOG(FLUID_WARN, "Sample '%s': invalid start/end file positions", sample->name);
+        return FLUID_FAILED;
+    }
+
+    return TRUE;
+}
+
+/* Check the sample loop pointers and optionally convert them to something
+ * usable in case they are broken. Return a boolean indicating if the pointers
+ * have been modified, so the user can be notified of possible audio glitches.
+ */
+int fluid_sample_sanitize_loop(fluid_sample_t *sample, unsigned int buffer_size)
+{
+    int modified = FALSE;
+    unsigned int max_end = buffer_size / 2;
+    /* In fluid_sample_t the sample end pointer points to the last sample, not
+     * to the data word after the last sample. FIXME: why? */
+    unsigned int sample_end = sample->end + 1;
+
+    /* Checking loops on compressed samples makes no sense at all and is really
+     * a programming error. Disable the loop to be on the safe side. */
+    if (sample->sampletype & FLUID_SAMPLETYPE_OGG_VORBIS)
+    {
+        FLUID_LOG(FLUID_ERR, "Sample '%s': checking loop on compressed sample, disabling loop",
+                sample->name);
+        sample->loopstart = sample->loopend = 0;
+        return TRUE;
+    }
+
+    if (sample->loopstart == sample->loopend)
+    {
+        /* Some SoundFonts disable loops by setting loopstart = loopend. While
+         * technically invalid, we decided to accept those samples anyway. Just
+         * ensure that those two pointers are within the sampledata by setting
+         * them to 0. Don't set modified here, as this change has no audible
+         * effect. */
+        sample->loopstart = sample->loopend = 0;
+    }
+    else if (sample->loopstart > sample->loopend)
+    {
+        unsigned int tmp;
+
+        /* If loop start and end are reversed, try to swap them around and
+         * continue validation */
+        FLUID_LOG(FLUID_DBG, "Sample '%s': reversed loop pointers '%d' - '%d', trying to fix",
+                sample->name, sample->loopstart, sample->loopend);
+        tmp = sample->loopstart;
+        sample->loopstart = sample->loopend;
+        sample->loopend = tmp;
+        modified = TRUE;
+    }
+
+    /* The SoundFont 2.4 spec defines the loopstart index as the first sample
+     * point of the loop while loopend is the first point AFTER the last sample
+     * of the loop. However we cannot be sure whether any of loopend or end is
+     * correct. Hours of thinking through this have concluded that it would be
+     * best practice to mangle with loops as little as necessary by only making
+     * sure the pointers are within sample->start to max_end. Incorrect
+     * soundfont shall preferably fail loudly. */
+    if ((sample->loopstart < sample->start) || (sample->loopstart > max_end))
+    {
+        FLUID_LOG(FLUID_DBG, "Sample '%s': invalid loop start '%d', setting to sample start '%d'",
+                sample->name, sample->loopstart, sample->start);
+        sample->loopstart = sample->start;
+        modified = TRUE;
```

Comment:
This actually introduces a regression for #171. As the comment points out we should not modify this in case of `(sample->loopstart < sample->start)`, at least for SF2 files. This handling would be probably ok for SF3, as it does not share a big sample buffer like SF2.

Reply:
I don't think it introduces a regression. It's not quite what the comment says, but that's how it is implemented in master:
```
invalid_loopstart = (sam->loopstart < sam->start) || (sam->loopstart >= sam->loopend);
```

And when invalid_loopstart is set, loopstart is set to sample->start.

Reply:
Sry, my bad. I was convinced it sounded wrong during a first rehearsal. But you're right, it's fine.